README.md
Omoikane
========

Poderoso punto de venta opensource de tecnología de vanguardia. Solución sencilla para cualquier tipo de negocio, especializado en negocios, que por sus características complejas no pueden ser tratados con puntos de venta básicos. (la mayoría en el mercado). Puede manejar desde 1 producto hasta cientos de miles sin problemas. Tiene características de inventario y almacén sencillas de uso, pero avanzadas en características que superan la media del mercado.

![Caja](http://i.imgur.com/JHh7C0f.png)

## Características
![Alt MainMenu](https://github.com/Phesus/Omoikane/raw/gh-pages/images/SS-MainMenu-2013-06-20.png)
- __Opensource__. Cualquiera puede extender nuestro código para implementar o modificar funciones en sus negocios, puede ser un excelente punto de partida para negocios que por su tamaño o especialización requiern una solución propia. Ésto siempre que respeten los términos de nuestra licencia :).
- __Diseño__ moderno que da una buena presentación a los clientes y oscuro que reduce el cansancio visual de los usuarios
- __Tuchscreen y touchless (no touch)__: No solo su diseño es perfectamente apto para terminales touchscreen, además presenta una navegación por teclado, finalmente cuidada para terminales que no tienen touchscreen, ésto es útil en negocios de mucho tráfico.
- __Seguridad__ de transacciones: Una de las características más cuidadas ha sido la protección contra robo por parte del personal que labora en caja.
- __Huela dactilar__ Autenticación de usuarios por tecnología biométrica de reconocimiento de huellas dactilares. Por el momento solo disponible para lectores U.Are.U 4000 y 4500.

![Imgur](http://i.imgur.com/08IbsmI.png)
- Autenticación estándar usando nip. ( Se recomienda usar la autenticación biométrica de ser posible ).
- __Multi-código__ de producto. Cada producto puede tener asignado un código de barras (o interno) e ilimitados códigos secundarios.
- __Paquetes__ de productos. Es posible agrupar distintos productos y en distintas cantidades bajo un mismo nombre y código de producto. Además esa agrupación de productos puede tener un precio especial que no necesariamente es la suma de sus componentes.
- __Categorización__ de productos. Es posible categorizar un producto por línea y por grupo. 
- __Descuentos__ por producto, por grupo y por línea. 
- __Robusta y confiable__ base de datos libre y gratuita MySQL. 
- __Rendimiento, Escalabilidad y capacidad__ comprobados: El rendimiento no depende del tamaño del negocio, nuestra aplicación funciona en negocios con más de dos millones de registros y consideramos que puede soportar decenas de millones sin afectar nunca el rendimiento.
- __Multi-impuesto__ Es posible agregar cualquier cantidad de impuestos a cualquier producto. Por ejemplo IVA 0%, IVA 16%, IEPS 8%, IEPS 25%, etc. Es posible agregar cualquier impuesto que sea una proporción del valor del producto.
- __Facturación y notas de crédito CFDI__ La facturación electrónica se puede realizar mediante diversos proveedores de timbres, desde 30 centavos por timbre aún comprando un solo timbre.
- __Catálogo offline__ Consulte su catálogo de artículos aún sin conexión. Útil en caso de caídas del servidor o para llevar consigo la información.
- __Niveles de precio__ Con omoikane es posible definir un número ilimitado de niveles de precios, por ejemplo: Precio 1, Precio 2, Precio 3, etc. Los niveles de precios se pueden asignar a clientes específicos o genéricos.

## Características agendadas:
- Agrupación por departamento. Actualmente se puede usar la agrupación por líneas como suplemento.
- Multi-almacén. Estamos analizando de que manera implementar esta característica sin impactar la simplicidad del sistema.
- Descuentos por cantidad. Una solución en la versión actual es usar los paquetes para dar descuentos por cantidad. Es útil para negocios que venden pocos productos en diferentes cantidades.
Además se agendarán las características sugeridas por nuestros usuarios, previo análisis por nuestra parte.

## Instalación
### Requisitos mínimos
- Windows XP y superiores, MAC OSX 10.6 o superiores, Linux con Java 7 o superior.
- Java JRE 7u21
- Al menos 2 GB de RAM
- Procesador de 1 GHz o superior
### Características recomendadas del sistema
- Procesador de doble núcleo de 2 GHz o superior
- Conexión a internet
- Para el uso de huella dactilar: Windows XP o superior y lector de huellas DigitalPersona u.are.u 4000 o 4500
- Para multi-caja se requiere una red LAN y/o VPN

### Instrucciones

[Quickstart / Instalación mono-caja](https://github.com/Phesus/Omoikane/wiki/Instalacion#quickstart--instalaci%C3%B3n-r%C3%A1pida-mono-caja)

[Instalación multi-caja](https://github.com/Phesus/Omoikane/wiki/Instalacion#instalaci%C3%B3n-multi-caja)

## Configuración

[Leer acerca de la configuración en nuestro wiki/manual](https://github.com/Phesus/Omoikane/wiki/04_configuracion)

## Manual

[Documentación](https://github.com/Phesus/Omoikane/wiki)

## Licencia

Licenciado bajo GNUv2. Para más detalles ver el archivo LICENSE adjunto al código en éste repositorio.


Contacto
========
- [Página de Phesus](http://phesus.com)

Instrumentación
===============
Para los preocupados por el lado técnico, Omoikane utiliza las siguientes tecnologías libres:
- Java
- MySQL
- JavaFX y swing en los componentes más viejos
- JasperReports
- Spring
- Hibernate

Descargas
=========

[Área de descargas](https://github.com/Phesus/Omoikane/releases)


CHANGELOG.md
## Versión 4.4.1 (2015-10-07)
- Corregí el reporte de artículos por grupos
- Corregí el reporte de artículos por líneas
- Corregí la migración 5 (v 4_4) a la que le faltaba una coma en un statement
- Añadí un try-catch a la línea de captura de caja
- Corrección de bug que no devuelve el enfoque luego de cancelaciones al campo de captura. Moví a Platform.runLater el 
código para recuperar el enfoque luego de una cancelación parcial o completa,en lugar de intentarlo en el mismo evento 
en el que se termina la cancelación
- Reduje la duración del timer del buscador integrado en caja de 1000ms a 500ms
- Agregué la capacidad de marcar la fecha de pago de las compras, está protegida por el permiso PMA_MARCAR_COMPRA_PAGADA
- Cambié la localización de US a "es-MX"
- Agregué una validación a los códigos alternos para verificar que no haya códigos repetidos
- Optimicé (un poco) la apariencia y funcionamiento de la pestaña "Compras" dentro de la ventana "Detalles de artículo"
- Mejoré el formato de impresión del catálogo de artículos

## Versión 4.4.0 (2015-08-22)
- Resolví bug que permitía agregar códigos repetidos como alternos, ésto ocasionaba un problema al cobrar productos con códigos repetidos.
 En primera instancia multiplicaba los impuestos por la cantidad de códigos repetidos al hacer una venta, pero al intentar cobrar identificaba el problema y
  generaba una excepción, una solución era cerrando y abriendo la venta, así se corregía la cantidad y el problema, generando una venta correcta.
  El error sólo sucedía al pasar a la venta el código repetido, si el mismo producto tenía un código no-repetido entonces no se generaba
- Control de excepción en catálogo de artículos en la columna de stock
- Cambio del atributo folioOrigen de camel-case a snake-case: folio_origen (tabla Compra)
### Segundo build
- Agrandé horizontalmente la tabla de compras en ficha de Artículo
- Agregué la capacidad de establecer una fecha de pago para una compra

## Versión 4.4.0 rc3 (2015-08-15)
- #125 Desactivé la tecla ESC para descartar: inventarios, compras y traspasos
- #124 Arreglé bug que impedía visualizar movimientos de almacén
- Arreglé bug que insertaba el resultado de la búsqueda rápida en la venta, en lugar del contenido del campo de captura
- Mejoré plantilla de impresión del listado de movimientos de almacén

## Versión 4.4.0 rc2 (2015-08-14)
- #121 Agregué perfil de cancelación y auto-cancelación, supervisor y gerente ya no pueden auto-cancelar
- Agregué campo de folio al formato impreso de los traspasos
- Arreglé la búsqueda de caja que había quedado dañada por la actualización de java
- #118 Arreglé el panel de etiquetas, que por actualización de java no mostraba correctamente las celdas
- #120 Las cancelaciones no eran navegables por teclado por actualización de java
- #117 No funcionaba switcheo de cambio de cliente desde java 7u40
- Detalles de artículo no abría cuando no había anotaciones
- Grupo y línea mostraban la línea en detalles de artículo

## Versión 4.4.0 rc1 (2015-08-11)
- Optimización del tiempo de apertura de la ventana de detalles/modificar artículo. 4x más rápida, 3 consultas que realizaban una conexión cada una fueron sustituidas por una que utiliza una
conexión del pool 

## Versión 4.4.0 b14 (2015-08-10)
- Se corrige bug que permitía inyectar código SQL en el catálogo de artículos y generaba un error buscando productos con apóstrofes, comillas simples, acentos sin letra.

## Versión 4.4.0 b13 (2015-08-05)
- Agregué un nuevo método "getCostoNeto" a la lógico de precios, obtiene el costo con impuestos
- Ahora los traspasos salientes muestran el costo neto (con impuestos) en lugar del costo bruto
- Corregí bug que permitía borrar partidas en traspasos
- Mejoré la plantilla de traspasos salientes, agregando espacio para las firmas y corrigiendo el formato de los números

## Versión 4.4.0 b12 (2015-08-04)
- Corregí bug que mostraba el total de un traspaso de mercancía en base al precio al público en lugar del costo
- Corregí un bug, no se marcaban las compras como completadas

## Versión 4.4.0 b11 (2015-07-26)
- Los reportes básicos ahora apuntan a la configuración general del sistema y no a su propia clase
que leía directo del archivo y no trabajaba en conjunto con el proxy de sucursales

## Versión 4.4.0 b10 (2015-07-25)
- Quité la persistencia durante la edición de Compras, ahora toda la compra se guarda una 
única ocasión y se agregan stocks todo en una misma sesión
- Corregí los reportes básicos que quedaban pendientes: costeo por grupos y líneas y artículos por línea
- Corregí un bug que ocasionaba que no se pudieran agregar códigos alternos, siempre 
marcaba "código repetido"

## Versión 4.4.0 b9 (2015-07-15)
- Cambié el límite de pacientes listados simultáneamente para artemisa de 35 a 10 000
- Agregué un flush forzado a cada operación de compras
- Quité Cascade en la relación Many-to-Many Articulos-Impuestos

## Versión 4.4.0 b8 (2015-07-12)
- Se corrige el bug que no permitía generar etiquetas debido a un lazy initialization exception de impuestos.

## Versión 4.4.0 b7
- Se corrige el enfoque después de buscar productos mediante el catálogo (F1) en inventarios y compras, el enfoque se perdía ralentizando la captura
- Se agrega el acceso a traspasos entrantes y salientes

## Versión 4.4.0 b6 (2015-07-06)
- Se corrige la relación VentaDetalles - VentaDetallesImpuesto. Ahora VDI tiene una clave primaria compuesta por el ID de VentaDetalles y el ID del impuesto.

## Versión 4.3.0 (2014-08-02)

- Se agrega reporte remoto de excepciones mediante airbrake
- Departamentos, nuevo característica que da soporte para clasificación por productos por departamentos
- CRUD de departamentos en panel de configuración
- Panel de configuración redimensionable 
- NIP de usuario ahora acepta números con longitud de 4 a 8 dígitos
- Panel de confección de etiquetas ahora acepta códigos secundarios

main/src/omoikane/entities/Venta.java
package omoikane.entities;

import omoikane.clientes.Cliente;

import javax.persistence.*;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.sql.Date;
import java.sql.Time;
import java.sql.Timestamp;
import java.util.Arrays;
import java.util.Collection;

/**
 * Created by IntelliJ IDEA.
 * User: octavioruizcastillo
 * Date: 14/07/11
 * Time: 04:04
 * To change this template use File | Settings | File Templates.
 */
//@Entity
public class Venta {
                    private int id;

    @NotNull        private int usuarioId;

    @NotNull        private Timestamp modificado;

    @NotNull        private Timestamp creado;

    @Min(0) @Max(1) private int facturada;

    @Min(0) @Max(1) private int completada;

                    private byte[] eliminar;

    @Min(0)         private BigDecimal subtotal;

    @Min(0)         private BigDecimal descuento;

    @Min(0)         private BigDecimal impuestos;

    @Min(0)         private BigDecimal total;

    @Min(0)         private BigDecimal efectivo;

    @Min(0)         private BigDecimal cambio;

    @Min(0)         private long folio;

    @NotNull        private Date fecha;

    @NotNull        private Time hora;

    @Column(name = "id")
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    @Column(name = "usuario_id")
    @Basic
    public int getUsuarioId() {
        return usuarioId;
    }

    public void setUsuarioId(int usuarioId) {
        this.usuarioId = usuarioId;
    }

    @Column(name = "modificado")
    @Basic
    public Timestamp getModificado() {
        return modificado;
    }

    public void setModificado(Timestamp modificado) {
        this.modificado = modificado;
    }

    @Column(name = "creado")
    @Basic
    public Timestamp getCreado() {
        return creado;
    }

    public void setCreado(Timestamp creado) {
        this.creado = creado;
    }

    @Column(name = "facturada")
    @Basic
    public int getFacturada() {
        return facturada;
    }

    public void setFacturada(int facturada) {
        this.facturada = facturada;
    }

    @Column(name = "completada")
    @Basic
    public int getCompletada() {
        return completada;
    }

    public void setCompletada(int completada) {
        this.completada = completada;
    }

    @Column(name = "eliminar")
    @Basic
    public byte[] getEliminar() {
        return eliminar;
    }

    public void setEliminar(byte[] eliminar) {
        this.eliminar = eliminar;
    }

    @Column(name = "subtotal")
    @Basic
    public BigDecimal getSubtotal() {
        return subtotal;
    }

    public void setSubtotal(BigDecimal subtotal) {
        this.subtotal = subtotal;
    }

    @Column(name = "descuento")
    @Basic
    public BigDecimal getDescuento() {
        return descuento;
    }

    public void setDescuento(BigDecimal descuento) {
        this.descuento = descuento;
    }

    @Column(name = "impuestos")
    @Basic
    public BigDecimal getImpuestos() {
        return impuestos;
    }

    public void setImpuestos(BigDecimal impuestos) {
        this.impuestos = impuestos;
    }

    @Column(name = "total")
    @Basic
    public BigDecimal getTotal() {
        return total;
    }

    public void setTotal(BigDecimal total) {
        this.total = total;
    }

    @Column(name = "efectivo")
    @Basic
    public BigDecimal getEfectivo() {
        return efectivo;
    }

    public void setEfectivo(BigDecimal efectivo) {
        this.efectivo = efectivo;
    }

    @Column(name = "cambio")
    @Basic
    public BigDecimal getCambio() {
        return cambio;
    }

    public void setCambio(BigDecimal cambio) {
        this.cambio = cambio;
    }

    @Column(name = "folio")
    @Basic
    public long getFolio() {
        return folio;
    }

    public void setFolio(long folio) {
        this.folio = folio;
    }

    @Column(name = "fecha")
    @Basic
    public Date getFecha() {
        return fecha;
    }

    public void setFecha(Date fecha) {
        this.fecha = fecha;
    }

    @Column(name = "hora")
    @Basic
    public Time getHora() {
        return hora;
    }

    public void setHora(Time hora) {
        this.hora = hora;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Venta venta = (Venta) o;

        if (venta.cambio.compareTo(cambio) != 0) return false;
        if (completada != venta.completada) return false;
        if (venta.descuento.compareTo(descuento) != 0) return false;
        if (venta.efectivo.compareTo(efectivo) != 0) return false;
        if (facturada != venta.facturada) return false;
        if (folio != venta.folio) return false;
        if (venta.impuestos.compareTo(impuestos) != 0) return false;
        if (venta.subtotal.compareTo(subtotal) != 0) return false;
        if (venta.total.compareTo(total) != 0) return false;
        if (usuarioId != venta.usuarioId) return false;
        if (id != venta.id) return false;
        if (!Arrays.equals(eliminar, venta.eliminar)) return false;
        if (fecha != null ? !fecha.equals(venta.fecha) : venta.fecha != null) return false;
        if (hora != null ? !hora.equals(venta.hora) : venta.hora != null) return false;
        if (modificado != null ? !modificado.equals(venta.modificado) : venta.modificado != null)
            return false;

        return true;
    }

    private Caja cajaByCajaId;

    @ManyToOne
    public
    @JoinColumn(name = "caja_id", referencedColumnName = "id", insertable=false, updatable=false)
    Caja getCajaByCajaId() {
        return cajaByCajaId;
    }

    public void setCajaByCajaId(Caja cajaByCajaId) {
        this.cajaByCajaId = cajaByCajaId;
    }

    private Cliente clienteByClienteId;

    @ManyToOne
    public
    @JoinColumn(name = "cliente_id", referencedColumnName = "cliente_id")
    Cliente getClienteByClienteId() {
        return clienteByClienteId;
    }

    public void setClienteByClienteId(Cliente clienteByClienteId) {
        this.clienteByClienteId = clienteByClienteId;
    }

    private Corte corte;

    @ManyToOne
    public
    @JoinColumns({@JoinColumn(name = "caja_id", referencedColumnName = "id_caja"),
                  @JoinColumn(name = "corte_sucursal_id", referencedColumnName = "corte_sucursal_id")})
    Corte getCorte() {
        return corte;
    }

    public void setCorte(Corte corte) {
        this.corte = corte;
    }

    private Collection<VentaDetalle> ventaDetallesByVentaId;

    @OneToMany(mappedBy = "ventaByVentaId")
    public Collection<VentaDetalle> getVentaDetallesByVentaId() {
        return ventaDetallesByVentaId;
    }

    public void setVentaDetallesByVentaId(Collection<VentaDetalle> ventaDetallesByVentaId) {
        this.ventaDetallesByVentaId = ventaDetallesByVentaId;
    }

    private Collection<VentaEspecial> ventaEspecialsByVentaId;

    @OneToMany(mappedBy = "ventaByIdVenta")
    public Collection<VentaEspecial> getVentaEspecialsByVentaId() {
        return ventaEspecialsByVentaId;
    }

    public void setVentaEspecialsByVentaId(Collection<VentaEspecial> ventaEspecialsByVentaId) {
        this.ventaEspecialsByVentaId = ventaEspecialsByVentaId;
    }

    private VentaFacturada ventaFacturadaByVentaId;

    @OneToOne(mappedBy = "ventaByVentaId")
    public VentaFacturada getVentaFacturadaByVentaId() {
        return ventaFacturadaByVentaId;
    }

    public void setVentaFacturadaByVentaId(VentaFacturada ventaFacturadaByVentaId) {
        this.ventaFacturadaByVentaId = ventaFacturadaByVentaId;
    }
}
